src/slab.rs

```diff
@@ -452,12 +452,24 @@ impl<T> IdSlab<T> {
     }
 
     fn get_mut_or_tagged_slot(&mut self, id: Id<T>) -> Result<&mut T, Option<&mut TaggedSlot<T>>> {
-        match self.slots.get_mut(id.index as usize) {
-            Some(&mut TaggedSlot {
-                slot: Slot::Occupied { ref mut value },
+        let tagged_slot = self.slots.get_mut(id.index as usize).ok_or(None)?;
```

Comment:
It's sad to introduce the double runtime check, I think you can actually write this as:

```rust
        match self.slots.get_mut(id.index as usize) {
            Some(tagged_slot) => {
                if id.tag == tagged_slot.tag {
                    match tagged_slot.slot {
                        Slot::Occupied { ref mut value } => Ok(value),
                        _ => Err(Some(tagged_slot)),
                    }
                } else {
                    Err(Some(tagged_slot))
                }
            }
            _ => Err(None),
        }
```

works on my end. Seems like the problem is to do with the pattern guard somehow.

Reply:
Yeah, I definitely like this approach!

BTW, I think there would be no double runtime check because of compiler optimizations. Look at this small example with immutable references:
https://gcc.godbolt.org/z/G1G-q8
https://gcc.godbolt.org/z/AV6unM